File: javascript/2015/day04b/day04b.js
#!/usr/bin/env node

const crypto = require('crypto')
const fs = require('fs');
const path = require('path');
const process = require('process');

function usage() {
    const progname = path.basename(process.argv[1]);
    console.log(`usage: ${progname} <key>`);
    process.exit(1);
}

function processKey(key) {
    let n = 1;
    while ( true ) {
        const tryKey = key + n;
        const digest = crypto.createHash('md5').update(tryKey).digest('hex');
        if ( digest.startsWith('000000') ) {
            return n;
        }
        n++;
    }
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    const key = process.argv[2];
    const result = processKey(key);
    console.log(`result = ${result}`);
}

main()

File: ruby/2015/day04b/day04b.rb
#!/usr/bin/env ruby

require 'digest'

def usage
    progname = File.basename($0)
    puts "usage: #{progname} <key>"
    exit 1
end

def process(key)
    n = 1
    while true
        try_key = key + n.to_s
        digest = Digest::MD5.hexdigest(try_key)
        if digest.start_with?("000000")
            return n
        else
            n += 1
        end
    end
end

def main
    usage if ARGV.length < 1
    key = ARGV[0]
    result = process(key)
    puts "result = #{result}"
end

main if __FILE__ == $0
